File: packages/android-sdk/gateflow/src/main/java/com/gateflow/sdk/dependencies/DependencyContainer.kt
package com.gateflow.sdk.dependencies

import android.content.Context
import com.gateflow.sdk.GateFlowOptions
import com.gateflow.sdk.analytics.EventTracker
import com.gateflow.sdk.config.ConfigManager
import com.gateflow.sdk.identity.IdentityManager
import com.gateflow.sdk.network.Network
import com.gateflow.sdk.storage.LocalStorage

/** Container holding all SDK components. Used for testing and internal DI. */
data class DependencyContainer(
    val storage: LocalStorage,
    val network: Network,
    val identityManager: IdentityManager,
    val configManager: ConfigManager,
    val eventTracker: EventTracker,
) {
    companion object {
        fun make(
            context: Context,
            baseUrl: String,
            apiKey: String,
            options: GateFlowOptions = GateFlowOptions(),
        ): DependencyContainer {
            val storage = LocalStorage(context)
            val network = Network(baseUrl, apiKey)
            val identityManager = IdentityManager(storage)
            val configManager = ConfigManager(storage, network, options.cacheTTL)
            val eventTracker = EventTracker(
                storage = storage,
                network = network,
                userIdProvider = { identityManager.state.value.userId },
                batchSize = options.eventBatchSize,
                flushIntervalMs = options.eventFlushInterval,
            )
            return DependencyContainer(
                storage = storage,
                network = network,
                identityManager = identityManager,
                configManager = configManager,
                eventTracker = eventTracker,
            )
        }
    }
}

File: packages/ios-sdk/Sources/GateFlowKit/Dependencies/DependencyContainer.swift
import Foundation

/// Container holding all SDK components. Used for testing and internal DI.
struct DependencyContainer {
    let storage: UserDefaultsStorage
    let network: Network
    let identityManager: IdentityManager
    let configManager: ConfigManager
    let eventTracker: EventTracker

    static func make(
        baseUrl: String,
        apiKey: String,
        storage: UserDefaultsStorage = UserDefaultsStorage(),
        options: GateFlowOptions = GateFlowOptions()
    ) -> DependencyContainer {
        let network = Network(baseUrl: baseUrl, apiKey: apiKey)
        let identityManager = IdentityManager(storage: storage)
        let configManager = ConfigManager(storage: storage, network: network, cacheTTL: options.cacheTTL)
        let eventTracker = EventTracker(
            storage: storage,
            network: network,
            userIdProvider: { identityManager.userId },
            batchSize: options.eventBatchSize,
            flushInterval: options.eventFlushInterval
        )
        return DependencyContainer(
            storage: storage,
            network: network,
            identityManager: identityManager,
            configManager: configManager,
            eventTracker: eventTracker
        )
    }
}
